### PR TITLE
[Merged by Bors] - chore: add Zulip emoji merge delegate workflow

### PR DESCRIPTION
**Pull request by jcommelin**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: .github/workflows/zulip_emoji_merge_delegate.yaml
Comment (bryangingechen) [inline .github/workflows/zulip_emoji_merge_delegate.yaml:13]:
```suggestion
      uses: actions/setup-python@v5
```

File: .github/workflows/zulip_emoji_merge_delegate.yaml
Comment (bryangingechen) [inline .github/workflows/zulip_emoji_merge_delegate.yaml:29]: Could you factor this script out into a separate file?

(Also, I guess you may already be aware, but secrets.ZULIP_EMAIL doesn't currently exist, we're just using `github-mathlib4-bot@leanprover.zulipchat.com` directly in our other scripts.)

File: .github/workflows/zulip_emoji_merge_delegate.yaml
Comment (bryangingechen) [inline .github/workflows/zulip_emoji_merge_delegate.yaml:56]: Not sure if this is the best way to do this, but presumably we'll want to filter out linkifiers pointing to other repos.
```suggestion
        pr_pattern = re.compile(r'(^|\s)#(\d{1,5})')
```
There's also the `!4#XXXXX` linkifier, but I think only the triage bot uses it (and only in the titles of threads).

File: .github/workflows/zulip_emoji_merge_delegate.yaml
Comment (jcommelin) [reply .github/workflows/zulip_emoji_merge_delegate.yaml:29]: Yep, I'll work on this asap

**Comment by grunweg** [comment]
Quick note: you will need to document the new script in `scripts/README.md`; this is why the `lint-style` job is failing.

File: .github/workflows/zulip_emoji_merge_delegate.yaml
Comment (adomani) [inline .github/workflows/zulip_emoji_merge_delegate.yaml:13]: I don't understand a large part of this script, but does it actually use the repository, or does it just access mathlib information online?

File: .github/workflows/zulip_emoji_merge_delegate.yaml
Comment (adomani) [inline .github/workflows/zulip_emoji_merge_delegate.yaml:32]:
```suggestion
        python scripts/zulip_emoji_merge_delegate.py "$ZULIP_API_KEY" "$ZULIP_EMAIL" "$ZULIP_SITE" "$GITHUB_TOKEN"
```
I think that this is what actionlint is talking about.

File: .github/workflows/zulip_emoji_merge_delegate.yaml
Comment (adomani) [reply .github/workflows/zulip_emoji_merge_delegate.yaml:13]: Thinking about this again, the script itself should be checked out!

Maybe you can do add a `sparse-checkout` flag to the action and just download the script itself, if that is really all that is needed by the action.

File: scripts/zulip_emoji_merge_delegate.py
Comment (adomani) [inline scripts/zulip_emoji_merge_delegate.py:22]:
```suggestion
# Fetch the last 200 messages
```
Is this a better description of what this command does?  I am just making a guess reading the code.

File: scripts/zulip_emoji_merge_delegate.py
Comment (adomani) [inline scripts/zulip_emoji_merge_delegate.py:43]: This means that the bot will only react to messages that do not already have a reaction.  Would it be better to have the bot react *always*?  After all, when a maintainer merges a PR, they usually react with :merge: when they `bors merge` the PR, not when the PR is actually merged.  The bot reaction would be confirmation that the PR really got merged after all.

File: scripts/zulip_emoji_merge_delegate.py
Comment (jcommelin) [reply scripts/zulip_emoji_merge_delegate.py:22]: yeah, forgot to update the comment :facepalm:

File: scripts/zulip_emoji_merge_delegate.py
Comment (jcommelin) [reply scripts/zulip_emoji_merge_delegate.py:43]: actually, maybe I should check for the `ready-to-merge` label, instead of `[Merged by bors]` in the title.

Comment (jcommelin) [reply scripts/zulip_emoji_merge_delegate.py:43]: Wait, that answers a different question

Comment (adomani) [reply scripts/zulip_emoji_merge_delegate.py:43]: Ok, that at least makes the two checks more consistent with one another!  :smile:

Comment (jcommelin) [reply scripts/zulip_emoji_merge_delegate.py]: I pushed an update that answers your question as well

**Review by adomani — APPROVED** [review]
I like this very much!

I do not understand the subtleties of the python code, but, as far as I can tell, this is great!  My opinion is that this is ready to be tested!

I'll also try the

maintainer delegate

option from review!

File: scripts/README.md
Comment (grunweg) [inline scripts/README.md:100]: This description is now outdate: the number is wrong, and there's a third emoji --- right?

File: scripts/zulip_emoji_merge_delegate.py
Comment (grunweg) [inline scripts/zulip_emoji_merge_delegate.py:55]: Optional: you could also have this *remove* a "delegated" reaction, right?

File: scripts/zulip_emoji_merge_delegate.py
Comment (grunweg) [inline scripts/zulip_emoji_merge_delegate.py:8]: Refer to the workflow file for an example how these variables look? Right now, reading just that comment would leave me pretty clue-less.

**Review by grunweg — APPROVED** [review]
The logic seems sensible (but I don't know the mechanics of working with zulip). Three small comments/suggestions.

**Comment by fpvandoorn** [comment]
Let's try this!

bors merge

**Pull request closed without merging**
